src/hotspot/share/tsan/tsanOopMap.cpp

```diff
@@ -414,81 +247,86 @@ namespace TsanOopMapImpl {
-                                      &n_downward_moves, &min_low, &max_high);
-    }
-    if (moves.length() == 0) {
-      return;
-    }
+// This is called with TSAN_ONLY, as we want to always create the weak
+// OopStorage so the number matches with the 'weak_count' in oopStorageSet.hpp.
+void TsanOopMap::initialize_map() {
+  // No need to do register_num_dead_callback for concurrent work as we do
+  // TsanOopMapTable cleanup, i.e. removing entries for freed objects during
+  // GC by calling TsanOopMap::notify_tsan_for_freed_and_moved_objects() from
+  // WeakProcessor.
+  _weak_oop_storage = OopStorageSet::create_weak("Tsan weak OopStorage", mtInternal);
+  assert(_weak_oop_storage != NULL, "sanity");
+
+  TSAN_RUNTIME_ONLY(
+    _oop_map = new TsanOopMapTable();
+  );
+}
 
-    // Notifying TSan is straightforward when source and target regions
-    // do not overlap:
-    if (disjoint_regions) {
-      for (int i = 0; i < moves.length(); ++i) {
-        const PendingMove &m = moves.at(i);
-        __tsan_java_move(m.source_begin(), m.target_begin(), m.n_bytes);
-      }
-      return;
-    }
+void TsanOopMap::destroy() {
+  delete _oop_map;
+}
 
-    // Source and target ranges overlap, the moves need to be ordered to prevent
-    // overwriting. Overall, this can take N^2 steps if only one object can be
-    // moved during the array traversal; however, when we are dealing with
-    // compacting garbage collector, observation shows that the overwhelming
-    // majority of the objects move in one direction. If we sort the moves (in
-    // the ascending order if dominant direction is downwards, in the descending
-    // order otherwise), chances are we will be able to order the moves in a few
-    // traversals of the moves array.
-    {
-      moves.sort((2 * n_downward_moves > moves.length()) ? lessThan : moreThan);
-      log_debug(gc)("Tsan: sort %d objects", moves.length());
-    }
+OopStorage* TsanOopMap::oop_storage() {
+  assert(_weak_oop_storage != NULL, "sanity");
+  return _weak_oop_storage;
+}
 
-    {
-      handle_overlapping_moves(moves, min_low, max_high);
-    }
+// Called during GC by WeakProcessor.
+void TsanOopMap::notify_tsan_for_freed_and_moved_objects() {
+  if (_oop_map == nullptr) {
+    return;
   }
 
-}  // namespace TsanOopMapImpl
+  bool disjoint_regions;
+  int n_downward_moves = 0;
+  char *min_low, *max_high;
+  char *source_low = reinterpret_cast<char *>(UINTPTR_MAX);
+  char *source_high = NULL;
+  char *target_low = reinterpret_cast<char *>(UINTPTR_MAX);
+  char *target_high = NULL;
 
+  ResourceMark rm;
+  GrowableArray<TsanOopMapImpl::PendingMove> moves(MAX2((int)(_oop_map->size()), 100000));
 
-void TsanOopMap::initialize_map() {
-  TsanOopMapImpl::oop_map = new TsanOopMapImpl::TsanOopSizeMap(512);
-}
-
-void TsanOopMap::destroy() {
-  delete TsanOopMapImpl::oop_map;
-}
+  {
+    MutexLocker mu(TsanOopMap_lock, Mutex::_no_safepoint_check_flag);
```

Comment:
Do we really need this lock if it is called during a GC pause by only the VM thread? Currently it is called by multiple GC threads, so the lock might appear necessary.

Also see the comment in the old `TsanOopMap::weak_oops_do`. We should probably keep that comment.

Reply:
We need the lock here as it's done in concurrent worker threads that handle the WeakProcessor::WeakOopsDoTasks.

As discussed offline today, doing the tsanOopMap processing and tsan notification in WeakProcessor::Task::work is cleaner and more maintainable than other possible alternatives.